### PR TITLE
Provide instance to give a change to implement delegate methods of NSURLSession

### PR DESCRIPTION
**Pull request by ishkawa**
Users of APIKit may want to implement delegate methods of NSURLSession to hook events or to handle authentication challenge, so I think we have to provide instances for each subclasses of API and re-implement wrapper of NSURLSession using delegate methods (not block). Currently, an instance of API subclass is provided as a singleton because I would like to keep interface simple and multiple clients are not necessary for one web API.

@ikesyo how do you think?

### CODE REVIEW

**Comment by ikesyo** [comment]
I agree that this is a required feature, and the users could implement the delegate methods in API subclasses is quite simple usage.

File: APIKit/APIKit.swift [hunk 1]
```diff
@@ -43,6 +39,41 @@ public class API {
...
+    // prevent instantiation
+    override private init() {
+        super.init()
+    }
+    
+    // create session and instance of API for each subclasses
+    private final class var instancePair: (API, NSURLSession) {
+        struct Singleton {
+            static var dictionary = [String: (API, NSURLSession)]()
+        }
```
Comment (ikesyo) [inline APIKit/APIKit.swift]: How about using private, global scope variable until we can use stored class var in Swift 1.2?

```swift
private var instancePairDictionary = [String: (API, NSURLSession)]()

public class API {
}
```

File: APIKit/APIKit.swift [hunk 2]
```diff
@@ -43,6 +39,41 @@ public class API {
...
+        
+        let className = NSStringFromClass(self)
+        var pair: (API, NSURLSession) = Singleton.dictionary[className] ?? {
+            let instance = (self as NSObject.Type)() as API
```
Comment (ikesyo) [inline APIKit/APIKit.swift]: Can write simply `let instance = self.init()`

File: APIKit/APIKit.swift [hunk 3]
```diff
@@ -43,6 +39,41 @@ public class API {
...
+    }
+    
+    // create session and instance of API for each subclasses
+    private final class var instancePair: (API, NSURLSession) {
+        struct Singleton {
+            static var dictionary = [String: (API, NSURLSession)]()
+        }
+        
+        let className = NSStringFromClass(self)
+        var pair: (API, NSURLSession) = Singleton.dictionary[className] ?? {
```
Comment (ikesyo) [inline APIKit/APIKit.swift]: Can use `let`

File: APIKit/APIKit.swift [hunk 4]
```diff
@@ -26,16 +26,12 @@ public enum Method: String {
     case CONNECT = "CONNECT"
 }
 
-public class API {
+public class API: NSObject, NSURLSessionDelegate {
     // configurations
     public class func baseURL() -> NSURL {
         return NSURL()
```
Comment (ikesyo) [inline APIKit/APIKit.swift]: How about using `fatalError` function instead?

Comment (ikesyo) [inline APIKit/APIKit.swift]: (This is out of scope for this PR)

**Comment by ikesyo** [comment]
I added some minor notes. :eyeglasses: Altogether, looks good :sparkles:

File: APIKit/APIKit.swift [hunk 5]
```diff
@@ -43,6 +39,41 @@ public class API {
...
+        struct Singleton {
+            static var dictionary = [String: (API, NSURLSession)]()
+        }
+        
+        let className = NSStringFromClass(self)
+        var pair: (API, NSURLSession) = Singleton.dictionary[className] ?? {
+            let instance = (self as NSObject.Type)() as API
+            let configuration = self.URLSessionConfiguration()
+            let session = NSURLSession(configuration: configuration, delegate: instance, delegateQueue: nil)
+            Singleton.dictionary[className] = (instance, session)
```
Comment (ikesyo) [inline APIKit/APIKit.swift]: Should we use a lock or dispatching to private serial queue for thread safety?

File: APIKit/APIKit.swift (hunk 1 above)
Comment (ishkawa) [inline APIKit/APIKit.swift]: :ok_hand: 

I will fix it and add comment about class var in Swift 1.2.

File: APIKit/APIKit.swift (hunk 5 above)
Comment (ishkawa) [inline APIKit/APIKit.swift]: Yes, we should! I think locking is more suitable for this.

File: APIKit/APIKit.swift (hunk 4 above)
Comment (ishkawa) [inline APIKit/APIKit.swift]: What functionality does `fatalError` function replace?

**Comment by ishkawa** [comment]
Thank you for your reviewing! I will fix code and develop to support `NSURLSessionDelegate`.

**Comment by ishkawa** [comment]
Oops, neilpa/Assertions is no longer available :skull:  I replace it with ikesyo/Assertions swift-1.1.

File: APIKit/APIKit.swift (hunk 4 above)
Comment (ikesyo) [inline APIKit/APIKit.swift]: This class func is practically abstract (returning empty URL is meaningless). So if the user doesn't override it, it should immediately crash with a message IMO.

And bacase `fatalError` is `@noreturn` function, we don't need to return some value.

**Comment by ikesyo** [comment]
Thanks for picking it! I didn't report that :sweat:

**Comment by ishkawa** [comment]
No problem. I commit @3dec41cb78ea2e78b534147bab15e2a2565c3321 on master too.

File: APIKit/APIKit.swift (hunk 4 above)
Comment (ishkawa) [inline APIKit/APIKit.swift]: I got it. It's totally right :+1:

**Comment by ishkawa** [comment]
@ikesyo This is ready for review :sparkles:

File: APIKit/APIKit.swift [hunk 6]
```diff
@@ -119,4 +192,27 @@ public class API {
             return nil
         }
     }
+    
+    // MARK: - NSURLSessionDelegate
+    public func URLSession(session: NSURLSession, didReceiveChallenge challenge: NSURLAuthenticationChallenge, completionHandler: (NSURLSessionAuthChallengeDisposition, NSURLCredential!) -> Void) {
```
Comment (ikesyo) [inline APIKit/APIKit.swift]: This is an optional delegate method. Should we need to set up default implementation?

File: APIKit/APIKit.swift [hunk 7]
```diff
@@ -119,4 +192,27 @@ public class API {
...
+    // MARK: - NSURLSessionTaskDelegate
+    // TODO: add attributes like NS_REQUIRES_SUPER when it is available in future version of Swift.
+    public func URLSession(session: NSURLSession, task: NSURLSessionTask, didCompleteWithError connectionError: NSError?) {
+        if let dataTask = task as? NSURLSessionDataTask {
+            dataTask.completionHandler?(dataTask.responseBuffer, dataTask.response, connectionError)
+        }
+    }
+
+    // MARK: - NSURLSessionDataDelegate
+    public func URLSession(session: NSURLSession, dataTask: NSURLSessionDataTask, didReceiveResponse response: NSURLResponse, completionHandler: (NSURLSessionResponseDisposition) -> Void) {
```
Comment (ikesyo) [inline APIKit/APIKit.swift]: This is an optional delegate method. Should we need to set up default implementation? The document for this says:

> If you do not provide this delegate method, the session always allows the task to continue.

File: APIKit/APIKit.swift [hunk 8]
```diff
@@ -26,16 +26,45 @@ public enum Method: String {
...
+        if let responseBuffer = objc_getAssociatedObject(self, &dataTaskResponseBufferKey) as? NSMutableData {
+            return responseBuffer
+        } else {
+            let responseBuffer = NSMutableData()
+            objc_setAssociatedObject(self, &dataTaskResponseBufferKey, responseBuffer, UInt(OBJC_ASSOCIATION_RETAIN_NONATOMIC))
+            return responseBuffer
+        }
+    }
+    
+    private var completionHandler: ((NSData!, NSURLResponse!, NSError!) -> Void)? {
```
Comment (ikesyo) [inline APIKit/APIKit.swift]: I think this signature should be `((NSData, NSURLResponse?, NSError?) -> Void)?` to avoid using IUO.

File: README.md [hunk 9]
```diff
@@ -165,6 +165,26 @@ class GitHub: API {
...
+
+
+### NSURLSessionDelegate
+
+APIKit creates singleton instances for each subclasses of API and set them as delegates of NSURLSession,
+so you can add following features by implementing delegate methods.
+
+- Hook events of NSURLSession
+- Handle authentication challenges
+- Convert task to NSURLSessionDownloadTask
```
Comment (ikesyo) [inline README.md]: nitpicking: `a data task`

File: APIKitTests/APITests.swift [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        let expectation = expectationWithDescription("wait for response")
+        let request = MockAPI.Endpoint.Get()
+        
+        MockAPI.sendRequest(request) { response in
+            switch response {
+            case .Success:
+                XCTFail()
+                
+            case .Failure(let box):
+                assert(box.unbox.domain, ==, APIKitErrorDomain)
```
Comment (ikesyo) [inline APIKitTests/APITests.swift]: We can use `assertEqual` here.

File: APIKitTests/APITests.swift [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        let request = MockAPI.Endpoint.Get()
+        
+        MockAPI.sendRequest(request) { response in
+            switch response {
+            case .Success:
+                XCTFail()
+                
+            case .Failure(let box):
+                assertEqual(box.unbox.domain, error.domain)
+                assertEqual(box.unbox.code, error.code)
```
Comment (ikesyo) [inline APIKitTests/APITests.swift]: How about to extract error to variable as below?

**Comment by ikesyo** [comment]
Sorry for the late response :bow: I added some review notes. Could you check it out, and merge master into this?

File: APIKitTests/APITests.swift [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+        let request = MockAPI.Endpoint.Get()
+        
+        MockAPI.sendRequest(request) { response in
+            switch response {
+            case .Success:
+                XCTFail()
+                
+            case .Failure(let box):
+                assert(box.unbox.domain, ==, APIKitErrorDomain)
+                assertEqual(box.unbox.code, 400)
```
Comment (ikesyo) [inline APIKitTests/APITests.swift]: How about to extract error to variable as below?

File: APIKit/APIKit.swift (hunk 6 above)
Comment (ishkawa) [inline APIKit/APIKit.swift]: Ah, this was added by mistake. I will remove this.

**Comment by ishkawa** [comment]
Thank you for reviewing! I will merge this soon.

**Comment by ishkawa** [comment]
Oops. I overlooked 1 note. I'll fix it soon.

File: APIKit/APIKit.swift [hunk 13]
```diff
@@ -77,11 +148,13 @@ public class API {
...
     public class func sendRequest<T: Request>(request: T, handler: (Result<T.Response, NSError>) -> Void = {r in}) -> NSURLSessionDataTask? {
-        let session = URLSession()
+        let session = URLSession
         let mainQueue = dispatch_get_main_queue()
         
         if let URLRequest = request.URLRequest {
-            let task = session.dataTaskWithRequest(URLRequest) { data, URLResponse, connectionError in
+            let task = session.dataTaskWithRequest(URLRequest)
+            
+            task.completionHandler = { data, URLResponse, connectionError in
```
Comment (ikesyo) [inline APIKit/APIKit.swift]: When calling this handler, `dataTask.responseBuffer` isn't optional, so you can change `data` to non-optional :wink:

**Comment by ikesyo** [comment]
:tada:

File: APIKit/APIKit.swift (hunk 13 above)
Comment (ishkawa) [inline APIKit/APIKit.swift]: Good point! I'll fix it in the next minor version.
